### PR TITLE
Basic Compiler Functionality

### PR DESCRIPTION
**Pull request by alexandramedway**
Changes:
We added a generator. Right now, this reads directly from the stdin and then outputs a file called out.py. 
We fixed the parser so that the grammar is correct for arithmetic and function calls (eg, print)
We changed the ast to reflect parser changes. 
Updated Makefile.

### CODE REVIEW

File: compiler/Makefile [hunk 1]
```diff
@@ -4,9 +4,10 @@
 OCAMLC=ocamlc
 OCAMLLEX=ocamllex
 OCAMLYACC=ocamlyacc
-OBJS=parser.cmo scanner.cmo
+OBJS=parser.cmo scanner.cmo generator.cmo
 
-default: $(OBJS)
```
Comment (akalicki) [inline compiler/Makefile]: Can we keep the `default` target for consistency with the other makefiles? Maybe

```makefile
default: generator

generator: $(OBJS)
    ocamlc -g -o generator $(OBJS)
```

File: compiler/sast.mli [hunk 2]
```diff
@@ -0,0 +1,13 @@
+(*
```
Comment (akalicki) [inline compiler/sast.mli]: let's leave this file out for this PR if we're not adding anything to it

File: compiler/Makefile [hunk 3]
```diff
@@ -25,12 +26,15 @@ parser.ml parser.mli: parser.mly
...
+generator.cmx: scanner.cmx parser.cmx ast.cmi
 parser.cmo: ast.cmi parser.cmi
 parser.cmx: ast.cmi parser.cmi
-parser.cmi: ast.cmi
 scanner.cmo: parser.cmi
 scanner.cmx: parser.cmx
-ast.cmi:
+ast.cmi: 
+parser.cmi: ast.cmi
+sast.cmi: ast.cmi
```
Comment (akalicki) [inline compiler/Makefile]: along the lines of below: sast.ml isn't being used for anything yet so let's leave it out

File: compiler/generator.ml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+open Printf
+
+let rec txt_of_expr expr = match expr with
+  | Int_lit(i) -> string_of_int(i)
+  | Float_lit(f) -> string_of_float(f)
+  | String_lit(s) -> sprintf "\"%s\"" s
+  | Id(e) -> e
+  | Binop(e1, op, e2) ->
+      sprintf "%s %s %s" (txt_of_expr e1) (txt_of_op op) (txt_of_expr e2)
+  | Unop(op, e) -> sprintf "%s %s" (txt_of_op op) (txt_of_expr e)
```
Comment (akalicki) [inline compiler/generator.ml]: we shouldn't have a space between the two strings here right? e.g. `Unop(Minus, 5)` should evaluate to `-5` not `- 5`

File: compiler/generator.ml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+and process_stmt stmt = match stmt with
+  | State(expr) -> sprintf "%s" (txt_of_expr expr);;
+
+(* Do Compilation - Temporary: will eventually be its own file *)
+let lexbuf = Lexing.from_channel stdin in
+let program = Parser.program Scanner.token lexbuf in
+let out_file = open_out "out.py" in
+fprintf out_file "%s" (process_stmt_list program []); close_out out_file
+
+
```
Comment (akalicki) [inline compiler/generator.ml]: nit: only one blank line at end of file

File: test/parser/parserize.ml [hunk 6]
```diff
@@ -15,10 +15,10 @@ let rec eval = function
       | Mod -> "Binop(" ^ v1 ^ ", Mod, " ^ v2 ^ ")"
       | Pow -> "Binop(" ^ v1 ^ ", Pow, " ^ v2 ^ ")"
 
-let rec eval_stmts acc = function 
+let rec eval_stmts acc = function
   | [] -> "[" ^ (String.concat " ; " acc) ^ "]"
   | stmt :: tl -> match stmt with
-    | Expr(e) ->
+    | State(e) ->
       let e1 = eval e in eval_stmts (("Expr(" ^ e1 ^ ")") :: acc) tl
```
Comment (akalicki) [inline test/parser/parserize.ml]: print `State` instead of `Expr` and update all parser tests to match

**Comment by akalicki** [comment]
Looks great! Left a few comments to fix before merging. Addresses #17.

File: compiler/generator.ml (hunk 4 above)
Comment (alexandramedway) [inline compiler/generator.ml]: Good catch

**Comment by akalicki** [comment]
amazing

**Pull request merged**